aws_lambda.py

```diff
@@ -318,6 +324,81 @@ def get_signed_source(server_info, change):
             }
 
 
+def compare_records(a, b):
+    """Compare two lists of records. Returns empty list if equal."""
+    b_by_id = {r["id"]: r for r in b}
+    diff = []
+    for ra in a:
+        rb = b_by_id.pop(ra["id"], None)
+        if rb is None:
+            diff.append(ra)
+    diff = diff.extend(b_by_id.values())
+    return diff
+
+
+@command
+def consistency_checks(event, **kwargs):
+    """Check the collections content and status consistency.
+    """
+    server_url = event["server"]
+    auth = os.getenv("AUTH")
+    if auth:
+        auth = tuple(auth.split(":", 1)) if ":" in auth else BearerTokenAuth(auth)
+
+    client = Client(server_url=server_url, auth=auth)
+
+    # List signed collection using capabilities.
+    info = client.server_info()
+    try:
+        resources = info["capabilities"]["signer"]["resources"]
+    except KeyError:
+        raise ValueError("No signer capabilities found. Run on *writer* server!")
+
+    signed_collections = {}
+    preview_buckets = set()
+    for resource in resources:
+        if resource["source"]["collection"] is not None:
+            signed_collections[(resource["destination"]["bucket"], resource["destination"]["collection"])] = resource
+        else:
+            signed_collections[(resource["destination"]["bucket"],)] = resource
+        preview_buckets.add(resource["preview"]["bucket"])
+
+    monitored = client.get_records(bucket="monitor", collection="changes", _sort="bucket,collection")
+    for entry in monitored:
+        bid = entry["bucket"]
+        cid = entry["collection"]
+
+        # Skip preview collections entries
+        if bid in preview_buckets:
+            continue
+
+        if (bid, cid) in signed_collections:
+            r = signed_collections[(bid, cid)]
+        elif (bid,) in signed_collections:
+            r = signed_collections[(bid,)]
+            r["source"]["collection"] = r["preview"]["collection"] = r["destination"]["collection"] = cid
+        else:
+            raise ValueError(f"Unknown signed collection {bid}/{cid}")
+
+        source_metadata = client.get_collection(bucket=r["source"]["bucket"], id=r["source"]["collection"])["data"]
+
+        if source_metadata["status"] == "to-review":
+            source_records = client.get_records(**r["source"])
+            preview_records = client.get_records(**r["preview"])
+            diff = compare_records(source_records, preview_records)
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
+
+        elif source_metadata["status"] == "signed":
+            preview_records = client.get_records(**r["preview"])
+            dest_records = client.get_records(**r["destination"])
+            diff = compare_records(preview_records, dest_records)
+            if diff:
+                raise ValueError(f"Inconsistency detected: {diff}")
```

Comment:
You might want to indicate the bucket/collection where this inconsistency happened.

Reply:
Also, I think you could check `r["source"]` against either `r["preview"]` or `r["destination"]`.